src/PhpSpec/CodeAnalysis/TokenizedTypeHintRewriter.php

```diff
@@ -99,13 +99,34 @@ private function stripTypeHints($tokens)
                     }
                     break;
                 case self::STATE_READING_CLASS:
-                    if ($this->tokenHasType($token, T_STRING) && !$this->currentClass) {
+                    if('{' == $token && $this->currentFunction) {
+                        $this->state = self::STATE_READING_FUNCTION_BODY;
+                        $this->currentBodyLevel = 1;
+                    }
+                    elseif ($this->tokenHasType($token, T_STRING) && !$this->currentClass) {
                         $this->currentClass = $token[1];
                     }
                     elseif($this->tokenHasType($token, T_FUNCTION)) {
                         $this->state = self::STATE_READING_FUNCTION;
                     }
                     break;
+                case self::STATE_READING_FUNCTION_BODY:
+                    if ('{' == $token) {
+                        $this->currentBodyLevel++;
+                    }
+                    elseif ('}' == $token) {
+                        $this->currentBodyLevel--;
+
+                        if ($this->currentBodyLevel == 0) {
+                            $this->currentFunction = '';
+                        }
+                    }
+
+                    if (!$this->currentFunction) {
+                        $this->state = self::STATE_READING_CLASS;
```

Comment:
this state change should happen in the ``$this->currentBodyLevel == 0`` condition IMO. This is where you know you are leaving the function